10-JavaScript/js-intro.md

```diff
@@ -0,0 +1,439 @@
+
+## The Roots of JS
+
+JavaScript was originally developed in **ten days** in May 1995 by Brendan Eich, while he was working for Netscape Communications Corporation ([Wikipedia](https://en.wikipedia.org/wiki/JavaScript))
+
+Although it was developed under the name Mocha, the language was officially called LiveScript when it first shipped in beta releases of Netscape Navigator 2.0 in September 1995. It was renamed JavaScript when it was deployed in the Netscape browser version 2.0B3.
+
+### JavaScript != Java
+
+The change of name from LiveScript to JavaScript roughly coincided with Netscape adding support for _Java_ technology in its Netscape Navigator Web browser. The final choice of name caused confusion, giving the impression that the language was a spin-off of the Java programming language, and the choice has been characterized as a marketing ploy by Netscape to give JavaScript the cachet of what was then the hot new Web programming language.
+
+## Node is the name of the game
+
+[Node.js](https://nodejs.org/en/) is an open-source, cross-platform **runtime environment** for developing server-side Web applications. Node.js is not a JavaScript framework... It is a runtime environment that interprets JavaScript using Google's V8 JavaScript engine (same as what's in Chrome).
+
+1. install node using `$ brew install node`
+1. Verify `$ node --version` gives back a node version, and not an error.
+
+We can use Node to run JavaScript programs (`$ node my_program.js`) and use Node as a _REPL_ (`$ node`).
+
+Node is also packaged with __npm__, the _Node Package Manager_. npm packages are to Node like gems are to Ruby. We'll use npm to define, install, and manage dependencies in our Node applications.
+
+## JavaScript and Ruby are like cousins
+
+JavaScript and Ruby share the following similarities:
+
+- Ruby and JavaScript are both _interpreted_ languages
+- Both support object-oriented programming
+- Both are dynamically typed
+
+## Using the Node REPL
+
+> I want to learn a new programming language. The best way for me to learn it is by playing with it, seeing what I can do, and seeing what breaks it, in a safe and quick way. What tools can I use?
+
+> _-- Some cool programmer learning a new language_
+
+Instead of `irb` (the interactive ruby shell as a REPL for Ruby), we will use Node's REPL for playing around with JavaScript. Start the REPL with the terminal command `$ node`. You'll get a caret (`>`) prompt. From here, you can enter code and have it evaluated, just like using `irb` or the Rails console.
+
+Open the `node` REPL and put in the following commands. What do you get?
+
+1. `'Hello World!';`
+1. `2 + 1;`
+1. `100 * 4;`
+1. `2 + 'Hello World!';`
+1. `let x = 7;`
+1. `x + 2;`
+1. `typeof x;`
+1. `2 - 'Hello World!';`
+
+<details>
+
+  <summary>
+    Compare your answers here
+  </summary>
+
+  1. `'Hello World!'`
+  1. `3`
+  1. `400`
+  1. `'2Hello World!'`
+  1. `undefined`
+  1. `9`
+  1. `'number'`
+  1. `NaN`
```

Comment:
I think using source numbering (`1.`, `2.`) would be helpful in cases like this because it makes it easier to compare questions and answers and verify they are lined up correctly.

(Though I'm generally a proponent of numbering `<ol>` in Markdown since it I think it improves readability, and presumably the order is significant.)

Reply:
I may do this if I feel fussy enough to change all these numbers, I like the idea of explicitly numbering them

Reply:
(in my defense, it was a lot easier to do write it with generic `1. 1. 1.` etc while I was re-writing and re-ordering things)